File: CleverFerret/src/main/java/com/universalmedialibrary/data/repository/LibraryRepository.kt
package com.universalmedialibrary.data.repository

import com.universalmedialibrary.data.local.dao.LibraryDao
import com.universalmedialibrary.data.local.dao.LibraryScanSettingsDao
import com.universalmedialibrary.data.local.entity.Library
import com.universalmedialibrary.data.local.entity.LibraryScanSettings
import kotlinx.coroutines.flow.Flow
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Repository for library management operations
 * Provides abstraction layer between UI and data access
 */
@Singleton
class LibraryRepository @Inject constructor(
    private val libraryDao: LibraryDao,
    private val libraryScanSettingsDao: LibraryScanSettingsDao
) {

    fun getAllActiveLibraries(): Flow<List<Library>> = libraryDao.getAllActiveLibraries()

    fun getAllLibraries(): Flow<List<Library>> = libraryDao.getAllLibraries()

    suspend fun getLibraryById(libraryId: Long): Library? = libraryDao.getLibraryById(libraryId)

    suspend fun getLibraryByPath(path: String): Library? = libraryDao.getLibraryByPath(path)

    fun getLibrariesByType(type: String): Flow<List<Library>> =
        libraryDao.getLibrariesByTypeFlow(type)

    suspend fun createLibrary(library: Library): Long {
        val libraryId = libraryDao.insertLibrary(library)
        if (libraryId > 0) {
            val defaults = LibraryScanSettings.defaults(libraryId, library.path)
            libraryScanSettingsDao.upsert(defaults)
        }
        return libraryId
    }

    suspend fun updateLibrary(library: Library) = libraryDao.updateLibrary(library)

    suspend fun deleteLibrary(library: Library) = libraryDao.deleteLibrary(library)

    suspend fun toggleLibraryActive(libraryId: Long, isActive: Boolean) =
        libraryDao.setLibraryActive(libraryId, System.currentTimeMillis())

    suspend fun updateLastScanned(libraryId: Long, timestamp: Long = System.currentTimeMillis()) =
        libraryDao.updateLastScanned(libraryId, timestamp)

    suspend fun getActiveLibraryCount(): Int = libraryDao.getActiveLibraryCount()

    suspend fun getActiveLibraryCountByType(type: String): Int =
        libraryDao.getActiveLibraryCountByType(type)

    suspend fun validateLibraryPath(path: String): Boolean {
        // Basic path validation - could be enhanced with file system checks
        return path.isNotBlank() && getLibraryByPath(path) == null
    }

    /**
     * Get or create the default unified library
     * This is the primary library that contains all media types.
     * Uses @Synchronized to prevent race conditions during concurrent calls.
     */
    @Synchronized
    suspend fun getOrCreateDefaultLibrary(): Library {
        // Try to find an existing default library by name (not by hard-coded ID)
        val existingLibrary = libraryDao.getLibraryByName("My Library")
            ?: libraryDao.getLibraryByName("Universal Media Library")
        
        if (existingLibrary != null) {
            return existingLibrary
        }
        
        // Create the default unified library
        // Note: Uses "ALL" type which is a virtual library type that doesn't require scanning
        val defaultLibrary = Library(
            libraryId = 0,
            name = "My Library",
            type = "ALL", // Unified library supports all media types
            path = "", // Empty path - this is a virtual unified library, not a file-based one
            source = "VIRTUAL", // Mark as virtual to distinguish from file-based libraries
            description = "Your unified media library",
            isActive = true
        )
        
        val id = libraryDao.insertLibrary(defaultLibrary)
        val library = defaultLibrary.copy(libraryId = id)
        
        // Initialize scan settings for consistency with createLibrary()
        val defaults = LibraryScanSettings.defaults(id, library.path)
        libraryScanSettingsDao.upsert(defaults)
        
        return library
    }
    
    /**
     * Get the default library ID, creating it if it doesn't exist
     */
    suspend fun getDefaultLibraryId(): Long {
        return getOrCreateDefaultLibrary().libraryId
    }

    suspend fun getLibraryStats(): Map<String, Int> {
        val totalLibraries = getActiveLibraryCount()
        val bookLibraries = getActiveLibraryCountByType("BOOK")
        val musicLibraries = getActiveLibraryCountByType("MUSIC")
        val movieLibraries = getActiveLibraryCountByType("MOVIE")
        val otherLibraries = totalLibraries - bookLibraries - musicLibraries - movieLibraries

        return mapOf(
            "total" to totalLibraries,
            "books" to bookLibraries,
            "music" to musicLibraries,
            "movies" to movieLibraries,
            "other" to otherLibraries
        )
    }

    suspend fun getScanSettings(libraryId: Long, fallbackPath: String = ""): LibraryScanSettings {
        val existing = libraryScanSettingsDao.getSettings(libraryId)
        if (existing != null) return existing
        val libraryPath = if (fallbackPath.isNotBlank()) {
            fallbackPath
        } else {
            libraryDao.getLibraryById(libraryId)?.path ?: ""
        }
        val defaults = LibraryScanSettings.defaults(libraryId, libraryPath)
        libraryScanSettingsDao.upsert(defaults)
        return defaults
    }

    fun observeScanSettings(libraryId: Long): Flow<LibraryScanSettings?> =
        libraryScanSettingsDao.observeSettings(libraryId)

    suspend fun updateScanSettings(settings: LibraryScanSettings) {
        libraryScanSettingsDao.upsert(
            settings.copy(lastUpdated = System.currentTimeMillis())
        )
    }
}

File: pwa-demo/src/data/repository/LibraryRepository.ts
/**
 * Library Repository
 * 
 * Provides abstraction layer between UI and database for library operations.
 * Migrated from LibraryRepository.kt
 */

import { db } from '../../services/database-complete';
import type { Library } from '../local/entity';

export class LibraryRepository {
  /**
   * Get all active libraries
   */
  async getAllActiveLibraries(): Promise<Library[]> {
    return db.libraries
      .where('isActive')
      .equals(1)
      .sortBy('name');
  }

  /**
   * Get all libraries (active and inactive)
   */
  async getAllLibraries(): Promise<Library[]> {
    return db.libraries.orderBy('name').toArray();
  }

  /**
   * Get library by ID
   */
  async getLibraryById(libraryId: number): Promise<Library | undefined> {
    return db.libraries.get(libraryId);
  }

  /**
   * Get library by path
   */
  async getLibraryByPath(path: string): Promise<Library | undefined> {
    return db.libraries.where('path').equals(path).first();
  }

  /**
   * Get libraries by type
   */
  async getLibrariesByType(type: string): Promise<Library[]> {
    return db.libraries
      .where('type')
      .equals(type)
      .and(lib => lib.isActive === true)
      .toArray();
  }

  /**
   * Create new library
   */
  async createLibrary(library: Omit<Library, 'libraryId'>): Promise<number> {
    const newLibrary: Library = {
      ...library,
      libraryId: 0, // Will be auto-generated
      dateModified: Date.now(),
      isActive: true,
      lastScanned: undefined,
    };
    return db.libraries.add(newLibrary);
  }

  /**
   * Update existing library
   */
  async updateLibrary(library: Library): Promise<void> {
    await db.libraries.update(library.libraryId, {
      ...library,
      dateModified: Date.now(),
    });
  }

  /**
   * Delete library
   */
  async deleteLibrary(library: Library): Promise<void> {
    await db.libraries.delete(library.libraryId);
  }

  /**
   * Toggle library active status
   */
  async toggleLibraryActive(libraryId: number, isActive: boolean): Promise<void> {
    await db.libraries.update(libraryId, {
      isActive,
      dateModified: Date.now(),
    });
  }

  /**
   * Update last scanned timestamp
   */
  async updateLastScanned(libraryId: number, timestamp: number = Date.now()): Promise<void> {
    await db.libraries.update(libraryId, {
      lastScanned: timestamp,
    });
  }

  /**
   * Get count of active libraries
   */
  async getActiveLibraryCount(): Promise<number> {
    return db.libraries.where('isActive').equals(1).count();
  }

  /**
   * Get count of active libraries by type
   */
  async getActiveLibraryCountByType(type: string): Promise<number> {
    return db.libraries
      .where('type')
      .equals(type)
      .and(lib => lib.isActive === true)
      .count();
  }

  /**
   * Validate library path
   */
  async validateLibraryPath(path: string): Promise<boolean> {
    if (!path.trim()) return false;
    const existing = await this.getLibraryByPath(path);
    return existing === undefined;
  }

  /**
   * Get library statistics
   */
  async getLibraryStats(): Promise<LibraryStats> {
    const totalLibraries = await this.getActiveLibraryCount();
    const bookLibraries = await this.getActiveLibraryCountByType('BOOK');
    const musicLibraries = await this.getActiveLibraryCountByType('MUSIC');
    const movieLibraries = await this.getActiveLibraryCountByType('MOVIE');
    const otherLibraries = totalLibraries - bookLibraries - musicLibraries - movieLibraries;

    return {
      total: totalLibraries,
      books: bookLibraries,
      music: musicLibraries,
      movies: movieLibraries,
      other: otherLibraries,
    };
  }

  /**
   * Subscribe to library changes (returns cleanup function)
   */
  observeAllLibraries(callback: (libraries: Library[]) => void): () => void {
    const subscription = db.libraries
      .orderBy('name')
      .toArray()
      .then(callback);

    // Set up live query subscription
    const observable = db.libraries.toCollection().toArray();
    
    // Return cleanup function
    return () => {
      // Dexie will handle cleanup automatically
    };
  }
}

/**
 * Library statistics interface
 */
export interface LibraryStats {
  total: number;
  books: number;
  music: number;
  movies: number;
  other: number;
}

// Export singleton instance
export const libraryRepository = new LibraryRepository();
